driver-core/src/main/java/com/datastax/driver/core/policies/ErrorAwarePolicy.java

```diff
@@ -0,0 +1,322 @@
+ * for the query plan.
+ * <p/>
+ * The policy follows the builder pattern to be created, the {@link Builder} class can be created with
+ * {@link #builder} method.
+ */
+public class ErrorAwarePolicy implements ChainableLoadBalancingPolicy {
+
+    private static final Logger logger = LoggerFactory.getLogger(ErrorAwarePolicy.class);
+
+    private final LoadBalancingPolicy childPolicy;
+
+    private final int retryPeriodMillis;
+
+    PerHostErrorTracker errorTracker;
+
+    private ErrorAwarePolicy(Builder builder) {
+        this.childPolicy = builder.childPolicy;
+        this.retryPeriodMillis = builder.retryPeriod;
+        this.errorTracker = new PerHostErrorTracker(builder.errorsThreshold, builder.errorsFilter);
+    }
+
+    @Override
+    public LoadBalancingPolicy getChildPolicy() {
+        return childPolicy;
+    }
+
+    @Override
+    public void init(Cluster cluster, Collection<Host> hosts) {
+        childPolicy.init(cluster, hosts);
+        cluster.register(this.errorTracker);
+    }
+
+    @Override
+    public HostDistance distance(Host host) {
+        return childPolicy.distance(host);
+    }
+
+    @Override
+    public Iterator<Host> newQueryPlan(String loggedKeyspace, Statement statement) {
+        final Iterator<Host> childQueryPlan = childPolicy.newQueryPlan(loggedKeyspace, statement);
+
+        return new AbstractIterator<Host>() {
+
+            @Override
+            protected Host computeNext() {
+                long currentTime = System.currentTimeMillis();
+                while (childQueryPlan.hasNext()) {
+                    Host host = childQueryPlan.next();
+
+                    Long excludedTime = errorTracker.getExclusions().get(host);
+                    if (excludedTime == null) {
+                        // No entry in exclusions map means host hasn't crossed the threshold yet.
+                        return host;
+                    }
+
+                    if (currentTime - excludedTime > retryPeriodMillis) {
+                        return host;
+                    }
+                }
+
+                return endOfData();
+            }
+        };
+    }
+
+    @Override
+    public void onAdd(Host host) {
+        childPolicy.onAdd(host);
+    }
+
+    @Override
+    public void onUp(Host host) {
+        childPolicy.onUp(host);
+    }
+
+    @Override
+    public void onDown(Host host) {
+        childPolicy.onDown(host);
+    }
+
+    @Override
+    public void onRemove(Host host) {
+        childPolicy.onRemove(host);
+    }
+
+    /**
+     * Creates a new error aware policy builder given the child policy
+     * that the resulting policy should wrap.
+     *
+     * @param childPolicy the load balancing policy to wrap with error
+     *                    awareness.
+     * @return the created builder.
+     */
+    public static Builder builder(LoadBalancingPolicy childPolicy) {
+        return new Builder(childPolicy);
```

Comment:
Not really a problem, but I noticed with our other builders we pass the parameters of the builder into the constructor of the Class being built instead of the builder itself, should we do that here for consistency or does it not really matter?